### PR TITLE
Feat/adaptive templates geometry classes

### PR DESCRIPTION
**Pull request by matthewAnsys**
Adaptive templates functionality

### CODE REVIEW

File: src/ansys/motorcad/core/geometry.py [hunk 1]
```diff
@@ -1,6 +1,400 @@
...
+            Line/arc entity class instance
+        """
+        self.entities.append(entity)
+
+    def insert_entity(self, index, entity):
+        """Insert entity to list of region entities at given index.
+
+        Parameters
+        ----------
+        index : Integer
```
Comment (jgsdavies) [inline src/ansys/motorcad/core/geometry.py:37]:
```suggestion
        index : int
```

File: src/ansys/motorcad/core/geometry.py [hunk 2]
```diff
@@ -1,6 +1,400 @@
...
+            Line/arc entity class instance
+        """
+        self.entities.insert(index, entity)
+
+    def insert_polyline(self, index, polyline):
+        """Insert polyline at given index, polyline can be made up of line/arc entities.
+
+        Parameters
+        ----------
+        index : Integer
```
Comment (jgsdavies) [inline src/ansys/motorcad/core/geometry.py:49]:
```suggestion
        index : int
```

File: tests/test_geometry.py [hunk 3]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
     mc.check_if_geometry_is_valid(1)
 
     mc.set_variable("Slot_Depth", save_slot_depth)
+
+
+def test_set_adaptive_parameter_value():
+    mc.set_adaptive_parameter_value("test_parameter", 100)
```
Comment (jgsdavies) [inline tests/test_geometry.py:61]: Can we check the dataStore entry with get_array_variable and assert these are the same.

Also would be good to have 2 cases to test overwriting and adding a new var.

File: tests/test_geometry.py [hunk 4]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+
+
+def test_get_adaptive_parameter_value():
+    mc.set_adaptive_parameter_value("test_parameter_1", 100)
```
Comment (jgsdavies) [inline tests/test_geometry.py:65]: Test case needed for getting a var that doesn't exist

File: tests/test_geometry.py [hunk 5]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+    assert region.name == "Stator"
+
+
+def test_set_region():
+    region = geometry.Region()
+    region.name = "testing_region"
+    region.colour = (0, 0, 255)
+    region.material = "Air"
+
+    mc.set_region(region)
```
Comment (jgsdavies) [inline tests/test_geometry.py:84]: retrieve with get_region and ensure this has worked

File: tests/test_geometry.py [hunk 6]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+
+
+def test_get_entities_between_poly_start_end():
```
Comment (jgsdavies) [inline tests/test_geometry.py:87]: Can you comment this. Not sure what the expected result is/what this is doing

File: tests/test_geometry.py [hunk 7]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+
+def test_save_adaptive_script():
+    mc.save_adaptive_script(get_dir_path() + r"\test_files\adaptive_templates_script.py")
+
+
+def test_region_add_entity():
+    region = geometry.Region()
+    region.add_entity(geometry.Line((0, 0), (1, 1)))
+
+    assert len(region.entities) == 1
```
Comment (jgsdavies) [inline tests/test_geometry.py:105]: Can you assert anything about the object you've added. Would also be good to get the arc case checked

File: tests/test_geometry.py [hunk 8]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+    region.insert_polyline(
+        0,
+        [
+            geometry.Line((0, 0), (1, 1)),
+            geometry.Line((1, 1), (1, 0)),
+            geometry.Line((1, 0), (0, 0)),
+        ],
+    )
+
+    assert len(region.entities) == 3
```
Comment (jgsdavies) [inline tests/test_geometry.py:129]: Better check?

File: tests/test_geometry.py [hunk 9]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+    test_region.entities = []
+
+    assert test_region._to_json() == raw_region
+
+
+def test_line_get_coordinate_from_percentage_distance():
+    line = geometry.Line((0, 0), (1, 1))
+
+    x, y = line.get_coordinate_from_percentage_distance(0, 0, 0.5)
+    assert (x, y) == (0.5, 0.5)
```
Comment (jgsdavies) [inline tests/test_geometry.py:211]: maybe worth having a test case where x/y are different. Otherwise could miss typo where we return in wrong order

**Review by jgsdavies — CHANGES_REQUESTED** [review]

**Comment by jgsdavies** [comment]
Also can you build docs with ```tox -e doc``` and verify that the docstrings you are writing look correct in html docs

File: tests/test_geometry.py (hunk 6 above)
Comment (jgsdavies) [reply tests/test_geometry.py:87]: Also looks like this test is failing

File: src/ansys/motorcad/core/methods/adaptive_geometry.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            Polyline using lines/arcs
+
+        Returns
+        -------
+        :Object List
+            List of Line/Arc objects
+        """
+        self.connection.ensure_version_at_least("2024.0")
+        raw_region = region._to_json()
+        raw_entities = _convert_entities_to_json(polyline)
```
Comment (jgsdavies) [inline src/ansys/motorcad/core/methods/adaptive_geometry.py:102]: ```_convert_entities_to_json``` assumes we always pass a list. If we just pass a single line/arc this fails. Should be handled in ```_convert_entities_to_json```

**Review by jgsdavies — CHANGES_REQUESTED** [review]

File: src/ansys/motorcad/core/geometry.py [hunk 11]
```diff
@@ -1,6 +1,400 @@
...
+        self.entities = []
+
+        # expect other properties to be implemented here including number duplications, material etc
+
+    def add_entity(self, entity):
+        """Add entity to list of region entities.
+
+        Parameters
+        ----------
+        entity : Line/Arc class
```
[the remark this answers is not in the document]
Comment (matthewAnsys) [reply src/ansys/motorcad/core/geometry.py:27]: Done

File: tests/test_geometry.py [hunk 12]
```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
...
+    test_region.area = 5.1
+    test_region.centroid = (0.0, 1.0)
+    test_region.region_coordinate = (0.0, 1.1)
+    test_region.duplications = 10
+    test_region.entities = []
+
+    region = geometry.Region()
+    region._from_json(raw_region)
+
+    assert region.name == test_region.name
```
[the remark this answers is not in the document]
Comment (matthewAnsys) [reply tests/test_geometry.py:172]: This does not work, pointing to different regions in memory

File: src/ansys/motorcad/core/geometry.py [hunk 13]
```diff
@@ -1,6 +1,400 @@
...
+                    ):
+                        self.entities.remove(entity)
+
+    # method to receive region from Motor-CAD and create python object
+    def _from_json(self, json):
+        """Convert class from json object.
+
+        Parameters
+        ----------
+        json: Json object
```
[the remark this answers is not in the document]
Comment (jgsdavies) [reply src/ansys/motorcad/core/geometry.py:97]: I think this should be dictionary?

File: src/ansys/motorcad/core/geometry.py [hunk 14]
```diff
@@ -1,6 +1,400 @@
...
+            "centroid": {"x": self.centroid[0], "y": self.centroid[1]},
+            "region_coordinate": {"x": self.region_coordinate[0], "y": self.region_coordinate[1]},
+            "duplications": self.duplications,
+            "entities": _convert_entities_to_json(self.entities),
+        }
+
+        return region_dict
+
+
+class Line:
```
[the remark this answers is not in the document]
Comment (jgsdavies) [reply src/ansys/motorcad/core/geometry.py:158]: Not sure about using line as parent. Would rather have a common ancestor for both e.g. Entity class

File: src/ansys/motorcad/core/geometry.py [hunk 15]
```diff
@@ -1,6 +1,400 @@
...
+    """Python representation of Motor-CAD line entity."""
+
+    def __init__(self, start, end):
+        """Create line entity based upon start and end coordinates.
```
[the remark this answers is not in the document]
Comment (jgsdavies) [reply src/ansys/motorcad/core/geometry.py:162]: This hasn't been fixed

File: src/ansys/motorcad/core/geometry.py [hunk 16]
```diff
@@ -1,6 +1,400 @@
...
+
+        Parameters
+        ----------
+        start : (float, float)
```
[the remark this answers is not in the document]
Comment (jgsdavies) [reply src/ansys/motorcad/core/geometry.py:166]: Not addressed

File: src/ansys/motorcad/core/methods/adaptive_geometry.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+            Motor-CAD region object.
+        """
+        self.connection.ensure_version_at_least("2024.0")
+        raw_region = region._to_json()
+
+        method = "SetRegion"
+        params = [raw_region]
+        return self.connection.send_and_receive(method, params)
+
+    def check_closed_region(self, region):
```
Comment (jgsdavies) [inline src/ansys/motorcad/core/methods/adaptive_geometry.py:81]: Unused?

File: src/ansys/motorcad/core/methods/adaptive_geometry.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+        Parameters
+        ----------
+        region : ansys.motorcad.core.geometry.Region
+            Motor-CAD region object.
+        """
+        self.connection.ensure_version_at_least("2024.0")
+        pass
+
+    def check_collisions(self, region, regions_to_check):
```
Comment (jgsdavies) [inline src/ansys/motorcad/core/methods/adaptive_geometry.py:92]: Is it worth testing this function?

**Review by jgsdavies — CHANGES_REQUESTED** [review]

File: src/ansys/motorcad/core/methods/adaptive_geometry.py (hunk 17 above)
Comment (matthewAnsys) [reply src/ansys/motorcad/core/methods/adaptive_geometry.py:81]: Delphi side is waiting for reviewing. So currently not used but will be very soon

File: src/ansys/motorcad/core/methods/adaptive_geometry.py (hunk 18 above)
Comment (matthewAnsys) [reply src/ansys/motorcad/core/methods/adaptive_geometry.py:92]: Unable to test at the moment, waiting for review.

**Review by jgsdavies — APPROVED** [review]

**Pull request merged**